### PR TITLE
enhancement(remap): Enable casting timestamps into integers

### LINKED ISSUE #5730: The Remap `to_int` function should accept timestamps
Passing a timestamp to `to_int` should return the Unix representation of the time as an integer.

### PR DESCRIPTION
**Pull request by lucperkins**
Closes #5730.

Enables users to pass timestamps to the `to_int` function to get the Unix time elapsed rather than needing to use `format_timestamp` and pass in `%s` as the format. The discussion around this appears to be a bit unsettled so I'm happy to defer to others on the desired end state here.

### CODE REVIEW

**Comment by jszwedko** [comment]
As I mentioned in [the issue](https://github.com/timberio/vector/issues/5730#issuecomment-751735849) I'm still don't think, as a user, I'd think to look for `to_int` to serve this purpose. I'm also still wondering about other common integer timestamp formats including millisecond timestamps and nanosecond timestamps. I'm curious what others think though.

**Comment by binarylogic** [comment]
@jszwedko I don't see the downside in supporting this but agree that `to_unix` is clearer.

> common integer timestamp formats including millisecond timestamps and nanosecond timestamps

Are these common? Unix is defined as seconds, and in my experience, it's more common to use decimals to represent fractional seconds.

**Comment by jszwedko** [comment]
> @jszwedko I don't see the downside in supporting this but agree that `to_unix` is clearer.
> 
> > common integer timestamp formats including millisecond timestamps and nanosecond timestamps
> 
> Are these common? Unix is defined as seconds, and in my experience, it's more common to use decimals to represent fractional seconds.

This is just my anecdotal experience, but I see millisecond timestamps, at least, with some frequency. For example, I found the prometheus exporter uses them yesterday when investigating https://github.com/timberio/vector/issues/5744 . AWS uses them in many places of its API. I think they may be more common than using floats given precision issues.

I like `to_unix_timestamp` (and maybe `to_unix_timestamp_milli`).

I guess the only downside I can see is that I might, as a user, expect it to error rather than convert the timestamp, but this feels like it wouldn't be a terribly common case.

**Comment by jszwedko** [comment]
Here are the change results over 20 runs (first column is count):

```
     20 buffers/in-memory       none
     20 buffers/on-disk none
     20 complex/complex none
      1 create and insert array improved
     18 create and insert array none
      1 create and insert array regressed
     20 create and insert nested-keys   none
     20 create and insert single-level  none
     20 deserialize/"boo\\"p"   none
     20 deserialize/"boop"      none
     20 deserialize/"boop"."snoot"      none
     20 deserialize/foo[0].bar[0][0].baz        none
     20 deserialize/foo[0].bar[0].baz[0]        none
      1 deserialize/foo[0]      improved
     19 deserialize/foo[0]      none
     20 deserialize/foo.bar.baz.bat[0]  none
     20 deserialize/p4th_wi7h.numb3r5   none
     20 deserialize/regular."quoted"."quoted but spaces"."quoted.but.periods".lookup[0].nested_lookup[0][0]     none
     20 deserialize/simple_string       none
     20 downcase: literal_value none
     20 elasticsearch_indexes/dynamic   none
     20 elasticsearch_indexes/static    none
     20 files/files_without_partitions  none
     20 from_string/"boo\\"p"   none
     20 from_string/"boop"      none
     20 from_string/"boop"."snoot"      none
     20 from_string/foo[0].bar[0][0].baz        none
      1 from_string/foo[0].bar[0].baz[0]        improved
     18 from_string/foo[0].bar[0].baz[0]        none
      1 from_string/foo[0].bar[0].baz[0]        regressed
     20 from_string/foo[0]      none
     20 from_string/foo.bar.baz.bat[0]  none
     20 from_string/p4th_wi7h.numb3r5   none
     20 from_string/regular."quoted"."quoted but spaces"."quoted.but.periods".lookup[0].nested_lookup[0][0]     none
     20 from_string/simple_string       none
     20 http/compression/gzip(6)        none
     20 http/compression/none   none
     20 interconnected/interconnected   none
     20 isolated_buffers/channels/futures01     none
     20 isolated_buffers/channels/tokio none
     20 isolated_buffers/leveldb/both   none
     20 isolated_buffers/leveldb/reading        none
     20 isolated_buffers/leveldb/writing        none
     20 iterate all fields array        none
     20 iterate all fields nested-keys  none
     20 iterate all fields single-level none
     20 lua_add_fields/native   none
     19 lua_add_fields/v1       none
      1 lua_add_fields/v1       regressed
     19 lua_add_fields/v2       none
      1 lua_add_fields/v2       regressed
     20 lua_field_filter/native none
      1 lua_field_filter/v1     improved
     18 lua_field_filter/v1     none
      1 lua_field_filter/v1     regressed
      6 lua_field_filter/v2     improved
      8 lua_field_filter/v2     none
      6 lua_field_filter/v2     regressed
     20 parse_json: invalid_json_with_default   none
      2 parse_json: literal_value       improved
     16 parse_json: literal_value       none
      2 parse_json: literal_value       regressed
     20 partitioned_batching/batching_gzip(6)_2097152   none
     20 partitioned_batching/batching_gzip(6)_512000    none
      3 partitioned_batching/batching_none_2097152      improved
     15 partitioned_batching/batching_none_2097152      none
      2 partitioned_batching/batching_none_2097152      regressed
      3 partitioned_batching/batching_none_512000       improved
     17 partitioned_batching/batching_none_512000       none
     20 partitioned_batching/partitioned_batching_gzip(6)_2097152       none
     20 partitioned_batching/partitioned_batching_gzip(6)_512000        none
     20 partitioned_batching/partitioned_batching_none_2097152  none
     20 partitioned_batching/partitioned_batching_none_512000   none
     20 pipe/pipe_big_lines     none
     20 pipe/pipe_multiple_writers      none
     20 pipe/pipe_simple        none
     20 pipe/pipe_small_lines   none
      5 regex/regex     improved
      9 regex/regex     none
      6 regex/regex     regressed
     20 remap: add fields with add_fields       none
      1 remap: add fields with remap    improved
     18 remap: add fields with remap    none
      1 remap: add fields with remap    regressed
     20 remap: coerce with coercer      none
     20 remap: coerce with remap        none
     20 remap: parse JSON with json_parser      none
     20 remap: parse JSON with remap    none
     20 serialize/"boo\\"p"     none
     20 serialize/"boop"        none
     20 serialize/"boop"."snoot"        none
     20 serialize/foo[0].bar[0][0].baz  none
     20 serialize/foo[0].bar[0].baz[0]  none
     20 serialize/foo[0]        none
     20 serialize/foo.bar.baz.bat[0]    none
     20 serialize/p4th_wi7h.numb3r5     none
     20 serialize/regular."quoted"."quoted but spaces"."quoted.but.periods".lookup[0].nested_lookup[0][0]       none
     20 serialize/simple_string none
     20 to_string/"boo\\"p"     none
     20 to_string/"boop"        none
     19 to_string/"boop"."snoot"        none
      1 to_string/"boop"."snoot"        regressed
     19 to_string/foo[0].bar[0][0].baz  none
      1 to_string/foo[0].bar[0][0].baz  regressed
      1 to_string/foo[0].bar[0].baz[0]  improved
     19 to_string/foo[0].bar[0].baz[0]  none
     20 to_string/foo[0]        none
      1 to_string/foo.bar.baz.bat[0]    improved
     19 to_string/foo.bar.baz.bat[0]    none
     19 to_string/p4th_wi7h.numb3r5     none
      1 to_string/p4th_wi7h.numb3r5     regressed
     20 to_string/regular."quoted"."quoted but spaces"."quoted.but.periods".lookup[0].nested_lookup[0][0]       none
     20 to_string/simple_string none
     20 transforms/transforms   none
     20 upcase: literal_value   none
```

Or just the improved/regressed (out of 20 runs):

```
      1 create and insert array improved
      1 create and insert array regressed
      1 deserialize/foo[0]      improved
      1 from_string/foo[0].bar[0].baz[0]        improved
      1 from_string/foo[0].bar[0].baz[0]        regressed
      1 lua_add_fields/v1       regressed
      1 lua_add_fields/v2       regressed
      1 lua_field_filter/v1     improved
      1 lua_field_filter/v1     regressed
      6 lua_field_filter/v2     improved
      6 lua_field_filter/v2     regressed
      2 parse_json: literal_value       improved
      2 parse_json: literal_value       regressed
      3 partitioned_batching/batching_none_2097152      improved
      2 partitioned_batching/batching_none_2097152      regressed
      3 partitioned_batching/batching_none_512000       improved
      5 regex/regex     improved
      6 regex/regex     regressed
      1 remap: add fields with remap    improved
      1 remap: add fields with remap    regressed
      1 to_string/"boop"."snoot"        regressed
      1 to_string/foo[0].bar[0][0].baz  regressed
      1 to_string/foo[0].bar[0].baz[0]  improved
      1 to_string/foo.bar.baz.bat[0]    improved
      1 to_string/p4th_wi7h.numb3r5     regressed
```

Note that if we see a regression or improvement, we typically expect to see the reverse when it returns to the baseline unless the first baseline was the anomalous one.

The partitioned batching, `lua_field_filter/v2` and `regex` ones seem a bit noise-y yet. We will likely want to dig into those a bit more. Somewhat surprisingly the high level topology and disk buffer benchmarks were consistent (likely with substantial stdev though).

Observed improvements / regressions:

```
create and insert array +1.1537%        regressed
create and insert array -1.3325%        improved
deserialize/foo[0]      -1.2242%        improved
from_string/foo[0].bar[0].baz[0]        -4.8200%        improved
from_string/foo[0].bar[0].baz[0]        +5.1245%        regressed
lua_add_fields/v1       +1.2977%        regressed
lua_add_fields/v2       +1.3488%        regressed
lua_field_filter/v1     -1.4196%        improved
lua_field_filter/v1     +1.6972%        regressed
lua_field_filter/v2     +1.0509%        regressed
lua_field_filter/v2     +1.0938%        regressed
lua_field_filter/v2     -1.1647%        improved
lua_field_filter/v2     +1.3569%        regressed
lua_field_filter/v2     +1.7262%        regressed
lua_field_filter/v2     -1.8848%        improved
lua_field_filter/v2     -2.0932%        improved
lua_field_filter/v2     -2.1620%        improved
lua_field_filter/v2     +2.1798%        regressed
lua_field_filter/v2     -2.2583%        improved
lua_field_filter/v2     +3.1216%        regressed
lua_field_filter/v2     -3.6570%        improved
parse_json: literal_value       -1.2808%        improved
parse_json: literal_value       +1.2854%        regressed
parse_json: literal_value       -1.5913%        improved
parse_json: literal_value       +1.6182%        regressed
partitioned_batching/batching_none_2097152      -5.0137%        improved
partitioned_batching/batching_none_2097152      +5.5792%        regressed
partitioned_batching/batching_none_2097152      -6.4220%        improved
partitioned_batching/batching_none_2097152      -6.6038%        improved
partitioned_batching/batching_none_2097152      +7.2048%        regressed
partitioned_batching/batching_none_512000       -3.7311%        improved
partitioned_batching/batching_none_512000       -4.5033%        improved
partitioned_batching/batching_none_512000       -5.7850%        improved
regex/regex     +1.8791%        regressed
regex/regex     -1.9155%        improved
regex/regex     +2.0136%        regressed
regex/regex     -2.1610%        improved
regex/regex     +2.2352%        regressed
regex/regex     -2.4888%        improved
regex/regex     -2.6008%        improved
regex/regex     +2.7651%        regressed
regex/regex     -2.9066%        improved
regex/regex     +3.3316%        regressed
regex/regex     +6.0097%        regressed
remap: add fields with remap    -1.4700%        improved
remap: add fields with remap    +1.5678%        regressed
to_string/"boop"."snoot"        +1.2658%        regressed
to_string/foo[0].bar[0][0].baz  +1.9819%        regressed
to_string/foo[0].bar[0].baz[0]  -1.9835%        improved
to_string/foo.bar.baz.bat[0]    -2.1058%        improved
to_string/p4th_wi7h.numb3r5     +3.7470%        regressed
```

My recommendation is to try to take a closer look at the noise to see if there is a change we could make to reduce it, otherwise set the noise thresholds for partitioned batching to 10% and regex + lua_field_filter to 5%.

I'm going to try running multiple instances on the same dedicated host to ensure they won't be stepping on each other.

**Comment by spencergilbert** [comment]
The OpenTelemetry appears to specify unix epoch in nanoseconds, FWIW

**Comment by binarylogic** [comment]
Makes sense. We can also add a `to_unix(unit = seconds)` function. What do you think?

I still like what's going on in this PR though. `to_int(timestamp)` makes sense to me. Not that Ruby is the shining example of the perfect programming language, but Ruby allows for `DateTime.now.to_i`. I'm also not a fan of having multiple ways to do this. So I'll defer to @lucperkins on if we should close this PR.

**Comment by jszwedko** [comment]
:+1: yeah, I'm ok with including `to_int` support here if we also add `to_unix_timestamp` to allow converting to nano and milli second timestamps.

**Comment by jszwedko** [comment]
Just noting there is a failing test here: https://github.com/timberio/vector/pull/5753/checks?check_run_id=1620691258#step:8:898

**Comment by lucperkins** [comment]
@jszwedko Duly noted. I'll create an issue and carry that through in a separate PR.

**Comment by jamtur01** [comment]
@lucperkins This okay to be merged?

**Review by JeanMertz — APPROVED** [review]
Nice 👍

File: docs/reference/remap/to_int.cue [hunk 1]
```diff
@@ -2,9 +2,16 @@ remap: functions: to_int: {
...
 		{
 			name:        "value"
-			description: "The string that is to be converted to an int. Must be the string representation of an `int`, otherwise, an `ArgumentError` will be raised."
+			description: """
+				The value to convert to an integer.
+
+				* If a string, it must be the string representation of an integer or else an error
+					is raised.
+				* If a Boolean, returns `0` for `false` and `1` for `true`.
+				* If a timestamp, returns the [Unix timestamp](\(urls.unix_timestamp)).
```
Comment (binarylogic) [inline docs/reference/remap/to_int.cue:11]:
```suggestion
				* If a timestamp, returns the [Unix timestamp](\(urls.unix_timestamp)) in seconds.
```

**Pull request merged**